gravitational/teleport
# Report unexpected VNet shutdown from tsh daemon to Electron app

The main aim of this PR is to update the status of VNet in the Electron app when tsh daemon sees that VNet has unexpectedly stopped.

Just as with Connect My Computer, we don't generally expect VNet to unexpectedly stop. But when it happens, we want to make sure that the UI reflects the actual state of VNet.

https://github.com/gravitational/teleport/assets/27113/c438c3a2-bd58-44d1-8885-d1d34bc857a0

## Modeling VNet status

When working on Connect My Computer, we spent a lot of time on modelling the state of an agent correctly. CMC is much, much more complicated in that regard due to all the extra setup that's needed.

For VNet, I think I came up with a good mental model that's simpler. I'm going to use "status" and "state" interchangeably, though in the actual code we only use "status" because "state" is a very overloaded term in the context of React.

VNet can only be in two states: it's either working or not working. This gives us tangible feedback about how the system behaves. If VNet is running, all requests to apps are proxied through local proxies managed by VNet. If VNet is not running, then requests go through the proxy service as usual.

So there's just `'running'` and `'stopped'`. Notice how there's no place for `'starting'` or `'stopping'`. How do we represent that then?

There are only two states, but there are two actions that can transition VNet between those two states – starting and stopping. Those are represented by `useAsync` and its attempts. Actions can be not performed yet, in progress, or finished, either successfully or with an error.

(Well, technically there's a third action which is handled in this PR, an unexpected shutdown. But as this is not something performed by the app, there's no attempt that represents it.)

## How to review this

A good entry point to reviewing the JS changes is `vnetContext.tsx`. This is the "brain" of VNet in Connect, so it refers to all the other stuff I needed to adjust in this PR.

## How to test this

Testing an unexpected shutdown is simple, as you don't really need to use VNet for anything. We just need to start it and then kill the admin process.

Start VNet, then open Activity Monitor, View -> All Processes, Hierarchically, search for `tsh` and find `tsh` running under kernel_task -> authtrampoline -> bash -> tsh. Then just:

```sudo kill -s KILL <tsh pid>```

Doing so will leave leftovers in `/etc/resolver`, preventing your device from reaching the cluster until you start VNet. Usually just restarting VNet would help there, but I think this branch is a little bit outdated and that doesn't help. To manually clean this up, just removes files in that directory:

```
sudo rm -rf /etc/resolver/*
```

If you're testing VNet itself, make sure you're not doing this on a local cluster for which you have entries in `/etc/hosts`.
